dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverter.java

```diff
@@ -0,0 +1,153 @@
+package com.dynatrace.opentelemetry.metric;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.opentelemetry.api.metrics.common.Labels;
+import io.opentelemetry.sdk.metrics.data.DoublePointData;
+import io.opentelemetry.sdk.metrics.data.LongPointData;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The CumulativeToDeltaConverter provides a way to transform cumulative (total) values to delta
+ * values which can be ingested in Dynatrace.
+ */
+class CumulativeToDeltaConverter {
+  private final Cache<String, CacheValue> cache;
+
+  private static class CacheValue {
+    private final Number number;
+    private final long timeMillis;
+
+    public CacheValue(Number number, long timeMillis) {
+      this.number = number;
+      this.timeMillis = timeMillis;
+    }
+
+    public Number getNumber() {
+      return number;
+    }
+
+    public long getTimeMillis() {
+      return timeMillis;
+    }
+  }
+
+  /**
+   * Create a new converter.
+   *
+   * @param expireAfter The {@link Duration} after which cache entries will expire, calculated from
+   *     the last write operation.
+   */
+  CumulativeToDeltaConverter(Duration expireAfter) {
+    this.cache = CacheBuilder.newBuilder().expireAfterWrite(expireAfter).build();
+  }
+
+  private static String createIdentifier(String name, Labels labels, String type) {
+    return String.format("%s %s %s", name, getSortedLabelsString(labels), type);
+  }
+
+  /**
+   * Convert a total counter to a delta. The identity of a counter is defined by its name and its
+   * dimensions.
+   *
+   * @param metricName The name of the metric.
+   * @param point The {@link DoublePointData} object containing the data point.
+   * @return The delta to the previous value if there is one with the same identifier in the cache
+   *     or the value itself if there is not.
+   */
+  public Double convertDoubleTotalToDelta(String metricName, DoublePointData point) {
+    String identifier = createIdentifier(metricName, point.getLabels(), "DOUBLE");
+
+    double newValue = point.getValue();
+    // reset the map on nan or inf
+    if (Double.isNaN(newValue) || Double.isInfinite(newValue)) {
+      this.cache.invalidate(identifier);
+      return newValue;
+    }
+    final CacheValue cacheValue = this.cache.getIfPresent(identifier);
```

Comment:
It seems like the cache only expires old values when a new value is written. It could be the case that data is expired but still in the cache. Do you know if the cache will return the expired data or if it will return null?

Reply:
The javadoc for the `expireAfterWrite` method on the cache builder says:
```
Specifies that each entry should be automatically removed from the cache once a fixed duration has elapsed after the entry's creation, or the most recent replacement of its value.
...
Expired entries may be counted in Cache.size, but will never be visible to read or write operations. Expired entries are cleaned up as part of the routine maintenance described in the class javadoc.
```
`getIfPresent` returns null if the value is not present, so I assume this is working as intended.